feat: introduce API leveling, post_training to v1alpha

cdoern opened the pull request:
# What does this PR do?

Rather than have a single `LLAMA_STACK_VERSION`, we need to have a `_V1`, `_V1ALPHA`, and `_V1BETA` constant.

This also necessitated addition of `level` to the `WebMethod` so that routing can be handeled properly.

move post_training to `v1alpha` as it is under heavy development and not near its final state

implements leveling in #3317 

## Test Plan

existing tests should pass with newly generated schema. Conformance test will fail, this will need approval and ack from a maintainer as is put forward in the above document.

cdoern commented:
given that #3317 has merged, this is the next step!

cdoern commented:
note: integration tests will fail until the SDK is regenerated with `v1alpha/inference` as opposed to `v1/inference`

nathan-weinberg commented on docs/openapi_generator/generate.py:
```diff
@@ -25,7 +25,7 @@
 
 def str_presenter(dumper, data):
-    if data.startswith(f"/{LLAMA_STACK_API_VERSION}") or data.startswith(
+    if data.startswith(f"/{LLAMA_STACK_API_V1}") or data.startswith(f"/{LLAMA_STACK_API_V1ALPHA}") or data.startswith(f"/{LLAMA_STACK_API_V1BETA}") or data.startswith(
```
The second two options will never evaluate, as the first will always take precedent

cdoern replied:
true, thanks!

nathan-weinberg commented on docs/openapi_generator/pyopenapi/operations.py:
```diff
@@ -114,10 +114,14 @@ class EndpointOperation:
     def get_route(self) -> str:
+        # Get the API level from the webmethod decorator
+        webmethod = getattr(self.func_ref, "__webmethod__", None)
+        api_level = webmethod.level if webmethod and hasattr(webmethod, 'level') else LLAMA_STACK_API_V1
```
Hmmmm, do we want to assume `v1` if no level is specified? Perhaps `v1alpha` would be a better option?

nathan-weinberg commented on llama_stack/schema_utils.py:
```diff
@@ -13,6 +13,7 @@
 
 @dataclass
 class WebMethod:
+    level: str | None = "v1"
```
These should use the constants as well

nathan-weinberg reviewed (approved):
LGTM - a few unit tests might be good just to verify the requests to get levels work for expected scenarios (e.g. default of `v1` level if none is specified)

cdoern commented:
@mattf , @raghotham , any thoughts here?

cdoern commented:
@leseb @bbrowning @mattf I have made the following changes:

For each route that now has `v1alpha` I added a secondary `@webmethod` with the old `v1` route with `deprecated=True`.
This change necessitated some handling throughout the codebase for a list of web methods as opposed to a single one: for spec generation, for server standup, etc. When creating 

I also added a log in the `TracingMiddleware` when then web method is retrieved which warns the user if they are using one of these routes with `deprecated=True`.

These changes will allow us to use multiple web methods and `deprecated=` in the future when we change routes, move to new API versions, etc.


Please let me know if the changes make sense!

cdoern commented:
one final change: 

> there should be no default version, each webmethod should have to declare its version

I missed this yesterday so I just implemented this. Each route now declares alpha or v1 explicitly

cdoern commented on llama_stack/apis/agents/agents.py:
```diff
@@ -481,7 +482,7 @@ class Agents(Protocol):
     """
 
-    @webmethod(route="/agents", method="POST", descriptive_name="create_agent")
+    @webmethod(route="/agents", method="POST", descriptive_name="create_agent", level=LLAMA_STACK_API_V1)
```
@wukaixingxp v1 is the default, and agents was v1. It's set to v1 here because I didn't want to change the level due to the upcoming deprecation. 

That being said there are still active users of this API and I think there is a migration doc being written to help users migrate to responses.
